ui/app/components/app/permission-page-container/permission-page-container.component.js

```diff
@@ -0,0 +1,157 @@
+    rejectPermissionsRequest: PropTypes.func.isRequired,
+    selectedIdentity: PropTypes.object.isRequired,
+    permissionsDescriptions: PropTypes.object.isRequired,
+    domainMetadata: PropTypes.object.isRequired,
+    request: PropTypes.object,
+    redirect: PropTypes.bool,
+    permissionRejected: PropTypes.bool,
+  };
+
+  static defaultProps = {
+    redirect: null,
+    permissionRejected: null,
+    request: {},
+  };
+
+  static contextTypes = {
+    t: PropTypes.func,
+    metricsEvent: PropTypes.func,
+  };
+
+  constructor (props) {
+    super(props)
+    this.state = {
+      selectedPermissions: this.getRequestedMethodState(
+        this.getRequestedMethodNames(props)
+      ),
+      selectedAccount: props.selectedIdentity,
+    }
+  }
+
+  componentDidUpdate () {
+    const newMethodNames = this.getRequestedMethodNames(this.props)
+
+    if (!deepEqual(Object.keys(this.state.selectedPermissions), newMethodNames)) {
+      // this should be a new request, so just overwrite
+      this.setState({
+        selectedPermissions: this.getRequestedMethodState(newMethodNames),
+      })
+    }
+  }
+
+  getRequestedMethodState (methodNames) {
+    return methodNames.reduce(
+      (acc, methodName) => {
+        acc[methodName] = true
+        return acc
+      },
+      {}
+    )
+  }
+
+  getRequestedMethodNames (props) {
+    return Object.keys(props.request.permissions)
+  }
+
+  onPermissionToggle = methodName => () => {
+    this.setState({
+      selectedPermissions: {
+        ...this.state.selectedPermissions,
+        [methodName]: !this.state.selectedPermissions[methodName],
+      },
+    })
+  }
+
+  componentDidMount () {
+    this.context.metricsEvent({
+      eventOpts: {
+        category: 'Auth',
+        action: 'Connect',
+        name: 'Popup Opened',
+      },
+    })
+  }
+
+  onCancel = () => {
+    const { request, rejectPermissionsRequest } = this.props
+    rejectPermissionsRequest(request.metadata.id)
+  }
+
+  onSubmit = () => {
+    // sanity validation
+    if (!this.state.selectedAccount) {
+      throw new Error(
+        'Fatal: no account selected'
+      )
+    }
+
+    const {
+      request: _request, approvePermissionsRequest, rejectPermissionsRequest,
+    } = this.props
+
+    const request = {
+      ..._request,
+      permissions: { ..._request.permissions },
+    }
+
+    Object.keys(this.state.selectedPermissions).forEach(key => {
+      if (!this.state.selectedPermissions[key]) {
+        delete request.permissions[key]
+      }
+    })
+
+    if (Object.keys(request.permissions).length > 0) {
+      approvePermissionsRequest(request, [this.state.selectedAccount.address])
+    } else {
+      rejectPermissionsRequest(request.metadata.id)
+    }
+  }
+
+  render () {
+    const { request, permissionsDescriptions, domainMetadata, selectedIdentity, redirect, permissionRejected } = this.props
+
+    const requestMetadata = request.metadata
+
```

Comment:
This logic from lines 123-129 maybe could be somewhere other than the `render()` method... not sure if I care enough about this to change it...

Reply:
oh, well, it would cleanly fit into the container

Reply:
- [ ] move this logic into the container